### ISSUE TITLE
Feature request: automatically add nested defaults

### ISSUE THREAD

**Issue by theahura** [horejsek/python-fastjsonschema#178]
If I have a jsonschema like:
```
schema = {
    "type": "object",
    "properties": {
        "foo": {
            "type": "object",
            "properties": {
                "bar": {
                    "type": "string",
                    "default": "bar"
                }
            }
        },
    },
    "required": [],
    "additionalProperties": False
}
```
the output of validation will be 
```
{}
```
instead of 
```
{
  foo: {
    bar: "bar"
  }
}
```
is there a way to make it so that nested defaults are automatically populated?

**Comment by horejsek** [comment]
Not automatically, but you can always define default on the foo object as well, that is:

```
schema = {
    "type": "object",
    "properties": {
        "foo": {
            "type": "object",
            "default": {"bar": "bar"},
            "properties": {
                "bar": {
                    "type": "string",
                    "default": "bar"
                }
            }
        },
    },
    "required": [],
    "additionalProperties": False
}
```

Doing this automatically might be quite expensive call, and also surprising. Sometimes you might not want the default to be set automatically for foo, but once foo is defined, you want to just ensure it contains required properties.

Would this suggestion work for you?

If not, I might have an idea to provide some special value (fastjsonschema.NESTED_DEFAULT, for example), that could indicate that an algorithm should make an object and fill child defaults too. But it deviates from the specification, so would rather avoid it unless necessary.

**Comment by theahura** [comment]
I ended up doing what you suggested, but I don't love it as a solution. I think it violates DRY -- if we update the underlying schema for bar, we also need to update foo, and this may be easy to forget for more complicated schema types. That said, I agree that the spec is not fleshed out here -- anything related to default behavior feels surprising and unexpected for some set of usecases

**Comment by horejsek** [comment]
I agree that it can violite DRY. But at the same time it is more explicit I would say.

What would you say on my suggestion above? It could look like this:

```
schema = {
    "type": "object",
    "properties": {
        "foo": {
            "type": "object",
            "default": {"bar": fastjsonschema.NESTED_DEFAULT},
            "properties": {
                "bar": {
                    "type": "string",
                    "default": "bar"
                }
            }
        },
    },
    "required": [],
    "additionalProperties": False
}
```

**Comment by eike-fokken** [comment]
While it would be very expensive, maybe it would be nice, if fastjsonschema provides an additional function `add_defaults_recursively` (as you can see I am just GREAT at naming things...), that recursively adds defaults but must be called explicitly by the user so they know what they are signing up for.

In case you, @horejsek are interested, I could cook up a PR, but probably not before next week.

**Comment by eike-fokken** [comment]
Also: I just realized: If I have an array of objects ant the object contains properties with defaults, these ARE added.

Is the general rule the following?

"The topmost layer of defaults is added"?

**Comment by horejsek** [comment]
@eike-fokken To be honest, I'm not following what you mean?

**Comment by efokken-abb** [comment]
Now I worked again on this and now I feel I understand the behavior of the library and the behavior requested by this issue.

From my testing, fastjsonschema goes through all nesting from top to bottom and if at any one point a property is

listed AND has a default value AND is absent,

then the default will be inserted. This is independent of the nesting level of the json (above the defaulted value).

The OP wants the following: If an outer container does not have a default, but one of its properties does, than the outer container should automatically get a default that contains the default of the (inner) property.

Maybe this helps future readers. I can say that I don't need this functionality.
